hrm-domain/hrm-service/migrations/20230926214506-create-client-profiles-tables.js

```diff
@@ -0,0 +1,317 @@
+    `);
+    console.log(
+      "Index 'fki_ProfilesToIdentifiers_identifierId_accountSid_Identifiers_id_accountSid_fk' created.",
+    );
+
+    // Add existing identifiers to Identifiers table
+    await queryInterface.sequelize.query(`
+      DO $$    
+        DECLARE identifier public."Identifiers";
+        DECLARE profile public."Profiles";
+      BEGIN
+        -- Insert one identifier per distinct numnber-accountSid pair
+        INSERT INTO "Identifiers" ("identifier", "accountSid", "createdAt", "updatedAt")
+        SELECT DISTINCT "number" as "identifier", "accountSid", current_timestamp as "createdAt", current_timestamp as "updatedAt" 
+        FROM "Contacts" WHERE "number" IS NOT NULL;
+        
+        -- For each new Identifier
+        FOR identifier IN (SELECT * FROM "Identifiers") LOOP
+          -- Create a "blank" Profile
+          INSERT INTO "Profiles"("name", "accountSid", "createdAt", "updatedAt")
+          VALUES (NULL, identifier."accountSid", current_timestamp, current_timestamp)
+          RETURNING * INTO profile;
+        
+          -- Link the profile to the identifier
+          INSERT INTO "ProfilesToIdentifiers"("profileId", "identifierId", "accountSid", "createdAt", "updatedAt")
+          VALUES (profile.id, identifier.id, identifier."accountSid", current_timestamp, current_timestamp);
+        
+        END LOOP;
+      END$$  
+    `);
+    console.log('Profiles, Identifiers and relations populated');
+
+    // Add profileId and identifierId columns to Contacts table, and create indexes on them
+    await queryInterface.sequelize.query(`
+      ALTER TABLE IF EXISTS public."Contacts" ADD COLUMN IF NOT EXISTS "profileId" integer;
+
+      ALTER TABLE IF EXISTS public."Contacts" ADD CONSTRAINT "Contacts_profileId_Profiles_id_fk"
+      FOREIGN KEY ("profileId", "accountSid") REFERENCES public."Profiles" (id, "accountSid") MATCH SIMPLE ON UPDATE CASCADE ON DELETE SET NULL;
+      
+      CREATE INDEX IF NOT EXISTS "Contacts_profileId_accountSid" ON public."Contacts" USING btree
+      ("profileId" ASC NULLS LAST, "accountSid" COLLATE pg_catalog."default" ASC NULLS LAST);
+
+      ALTER TABLE IF EXISTS public."Contacts" ADD COLUMN IF NOT EXISTS "identifierId" integer;
+      
+      ALTER TABLE IF EXISTS public."Contacts" ADD CONSTRAINT "Contacts_identifierId_Identifiers_id_fk" 
+      FOREIGN KEY ("identifierId", "accountSid") REFERENCES public."Identifiers" (id, "accountSid") MATCH SIMPLE ON UPDATE CASCADE ON DELETE SET NULL;
+
+      CREATE INDEX IF NOT EXISTS "Contacts_identifierId_accountSid" ON public."Contacts" USING btree
+      ("identifierId" ASC NULLS LAST, "accountSid" COLLATE pg_catalog."default" ASC NULLS LAST)
+    `);
+
+    // Populate profileId and identifierId columns on Contacts
+    await queryInterface.sequelize.query(`
+      DO $$
+        DECLARE tmp RECORD;
+      BEGIN
+        FOR tmp IN (
```

Comment:
The `FROM` clause in the `UPDATE` statement can be used to update tables with data from other tables (or the same table with a self join): https://www.postgresql.org/docs/current/sql-update.html

In this case, instead of using a loop you could do this update in a single query, something like:

```
          UPDATE "Contacts" c SET "profileId" = tmp."profileId", "identifierId" = tmp."identifierId"
          FROM "ProfilesToIdentifiers" p2i
          LEFT JOIN "Identifiers" ids ON p2i."identifierId" = ids.id
          WHERE c."number" = ids."identifier" AND c."accountSid" = ids."accountSid";
```

Obvs loops are fine for migrations, but it's good practice to try to avoid them anyway, so you know how to tackle similar issues in places where you should be avoiding SQL loops if at all possible (i.e. production code).

Don't change this code if you have it working, just FYI for the future